components/form/form-errory-summary.js

```diff
@@ -0,0 +1,109 @@
+import '../alert/alert.js';
+import '../link/link.js';
+import '../expand-collapse/expand-collapse-content.js';
+import { css, html, LitElement } from 'lit-element/lit-element.js';
+import { bodyStandardStyles } from '../typography/styles.js';
+import { LocalizeCoreElement } from '../../lang/localize-core-element.js';
+import { RtlMixin } from '../../mixins/rtl-mixin.js';
+
+class FormErrorSummary extends LocalizeCoreElement(RtlMixin(LitElement)) {
+
+	static get properties() {
+		return {
+			errors: { type: Object, attribute: false },
+			_expanded: { type: Boolean, attribute: false }
+		};
+	}
+
+	static get styles() {
+		return [bodyStandardStyles, css`
+
+			:host {
+				display: block;
+			}
+
+			:host([hidden]) {
+				display: none;
+			}
+
+			.d2l-form-error-summary-header {
+				align-items: top;
+				cursor: pointer;
+				display: flex;
+				justify-content: space-between;
+				padding: 0.3rem 0.3rem 0.3rem 1.2rem;
+			}
+
+			:host([dir="rtl"]) .d2l-form-error-summary-header {
+				padding-left: 0.3rem;
+				padding-right: 1.2rem;
+			}
+
+			.d2l-form-error-summary-text {
+				align-items: center;
+				display: flex;
+			}
+
+			.d2l-form-error-summary-error-list {
+				margin: 0 0 0 1.2rem;
+				padding-bottom: 0.65rem;
+				padding-top: 0.3rem;
+			}
+
+			:host([dir="rtl"]) .d2l-form-error-summary-error-list {
+				margin-left: 0;
+				margin-right: 1.2rem;
+			}
+		`];
+	}
+
+	constructor() {
+		super();
+		this.errors = [];
+		this._expanded = true;
+	}
+
+	render() {
+		const errorSummaryTerm = this.errors.length === 1 ? 'components.form-error-summary.singleError' : 'components.form-error-summary.multipleErrors';
+		const errorSummary = html`
+			<d2l-alert type="critical" no-padding aria-live="polite">
+				<div class="d2l-form-error-summary-header" @click=${this._toggleExpandCollapse}>
+					<div class="d2l-form-error-summary-text">${this.localize(errorSummaryTerm, {count: this.errors.length})}</div>
+					<d2l-button-icon
+						aria-expanded=${this._expanded ? 'true' : 'false'}
+						icon=${this._expanded ? 'tier1:arrow-collapse-small' : 'tier1:arrow-expand-small' }
+						@click=${this._toggleExpandCollapse}>
+					</d2l-button-icon>
+				</div>
+				<d2l-expand-collapse-content ?expanded=${this._expanded}>
+					<ul class="d2l-form-error-summary-error-list">
+						${this.errors.map(error => html`
+							<li>
+								<d2l-link href=${error.href} @click=${error.onClick}>${error.message}</d2l-link>
+							</li>
+						`)}
+					</ul>
+				</d2l-expand-collapse-content>
+			</d2l-alert>
+		`;
+		return this.errors.length > 0 ? errorSummary : html``;
+	}
+
+	async focus() {
+		if (this.errors.length === 0) {
+			super.focus();
+			return;
+		}
+		await this.updateComplete;
+		await new Promise(resolve => requestAnimationFrame(() => requestAnimationFrame(resolve)));
```

Comment:
Eeek, seems pretty fragile eh. Were you able to see why the list items weren't available without delaying? I haven't seen us need to do this in other `focus()` methods.

Reply:
Yeah, I'm not super happy with it but it seems to be the same reason animations don't run when going from `display: none` to `display: block` without waiting for 2xRAF.

The reason its needed is because `focus` is called immediately after setting `errors` which can trigger's the conditional rendering. If it goes from being hidden due to having no errors to being visible then we need to:
1. Wait for `updateComplete` so that the `errors` has been updated.
1. Wait for the first `requestAnimationFrame` which runs just before the browser actually renders the new content.
1. Wait for the second `requestAnimationFrame` at which point we can guarantee the errors have rendered so the browser will properly focus on them.

Reply:
Hmm... could `d2l-expand-collapse-content-expand`'s promise be incorporated somehow?

Reply:
Also, what happens if it's collapsed?

Reply:
In this case the wait is needed because of the conditional render: `return this.errors.length > 0 ? errorSummary : html``;`. We can't use the expand event because it starts pre-expanded so there is no event fired.

Reply:
> Also, what happens if it's collapsed?

If its collapsed then there is no focus change. We could move it to the `expand-collapse` button if we want.

Reply:
I see -- that's complicated. I'd be curious what Jeff thinks RE: focus management if things are collapsed. Picturing situations where there are long forms and the summary isn't visible, so moving focus to it would bring it into view.

Reply:
Chatted with Jeff a bit and we're happy with moving focus to the expand button if the error list is collapsed.